perf/perf.go

```diff
@@ -0,0 +1,520 @@
+package perf
+
+import (
+	"encoding/binary"
+	"io"
+	"math"
+	"os"
+	"runtime"
+	"sync"
+	"sync/atomic"
+	"unsafe"
+
+	"github.com/cilium/ebpf"
+	"github.com/cilium/ebpf/internal"
+
+	"github.com/pkg/errors"
+	"golang.org/x/sys/unix"
+)
+
+type perfEventHeader struct {
+	Type uint32
+	Misc uint16
+	Size uint16
+}
+
+// perfEventRing is a page of metadata followed by
+// a variable number of pages which form a ring buffer.
+type perfEventRing struct {
+	fd     int
+	mmap   []byte
+	reader *ringReader
+}
+
+func newPerfEventRing(cpu int, opts ReaderOptions) (*perfEventRing, error) {
+	if opts.Watermark >= opts.PerCPUBuffer {
+		return nil, errors.Errorf("Watermark must be smaller than PerCPUBuffer")
+	}
+
+	// Round to nearest page boundary and allocate
+	// an extra page for meta data
+	pageSize := os.Getpagesize()
+	nPages := (opts.PerCPUBuffer + pageSize - 1) / pageSize
+	size := (1 + nPages) * pageSize
+
+	fd, err := createPerfEvent(cpu, opts.Watermark)
+	if err != nil {
+		return nil, errors.Wrap(err, "can't create perf event")
+	}
+
+	if err := unix.SetNonblock(fd, true); err != nil {
+		unix.Close(fd)
+		return nil, err
+	}
+
+	mmap, err := unix.Mmap(fd, 0, size, unix.PROT_READ|unix.PROT_WRITE, unix.MAP_SHARED)
+	if err != nil {
+		unix.Close(fd)
+		return nil, err
+	}
+
+	// This relies on the fact that we allocate an extra metadata page,
+	// and that the struct is smaller than an OS page.
+	// This use of unsafe.Pointer isn't explicitly sanctioned by the
+	// documentation, since a byte is smaller than sampledPerfEvent.
+	meta := (*unix.PerfEventMmapPage)(unsafe.Pointer(&mmap[0]))
+
+	ring := &perfEventRing{
+		fd:     fd,
+		mmap:   mmap,
+		reader: newRingReader(meta, mmap[meta.Data_offset:meta.Data_offset+meta.Data_size]),
+	}
+	runtime.SetFinalizer(ring, (*perfEventRing).Close)
+
+	return ring, nil
+}
+
+func createPerfEvent(cpu, watermark int) (int, error) {
+	attr := unix.PerfEventAttr{
+		Type:        unix.PERF_TYPE_SOFTWARE,
+		Config:      unix.PERF_COUNT_SW_BPF_OUTPUT,
+		Bits:        unix.PerfBitWatermark,
+		Sample_type: unix.PERF_SAMPLE_RAW,
+		Wakeup:      uint32(watermark),
```

Comment:
Setting it to 0 makes it fall back to the kernel behaviour, which uses bufsize / 2 as the watermark. The idea is that you can batch reading from the ring buffer by delaying the epoll notification until the buffer is at least half (or a value of your choosing) full. This ends up saving CPU at the cost of latency.

We set it to 1 if we want immediate feedback (so `xdpcap`) and to bufsize / 2 for our DoS mitigation packet sampler, where we want to save CPU. We've not benchmarked whether this really makes a difference.

We could probably get rid of this for now, and just hardcode watermark 1 or whatever.

Reply:
I think it's reasonable to specify via bytes. Easier to actually manage your application's memory footprint when the units are the standard units.

Reply:
Ah, I see. By all means if you have regular use cases for configuring this another way, then we can support that. I'm mainly coming from the perspective of making the default simple and stupid and to work more-or-less the way you'd expect it to. I think that as-is, the user *must* think about how many bytes of events to look for, instead of defaulting to process one event at a time unless you provide a different option.

Reply:
Yeah, I'll change this.